components/nimbus/src/lib.rs

```diff
@@ -210,25 +238,140 @@ impl NimbusClient {
+        Ok(res)
+    }
+
+    fn get_installation_date(&self, db: &Database, writer: &mut Writer) -> Result<DateTime<Utc>> {
+        // we first check our context
+        if let Some(context_installation_date) = self.app_context.installation_date {
+            let res = DateTime::<Utc>::from_utc(
+                NaiveDateTime::from_timestamp(context_installation_date / 1_000, 0),
+                Utc,
+            );
+            log::info!("[Nimbus] Retrieved date from Context: {}", res);
+            return Ok(res);
+        }
+        let store = db.get_store(StoreId::Meta);
+        let persisted_installation_date: Option<DateTime<Utc>> =
+            store.get(writer, DB_KEY_INSTALLATION_DATE)?;
+        let res = if let Some(installation_date) = persisted_installation_date {
+            installation_date
+        } else {
+            let installation_date = match self.get_creation_date_from_path(&self.db_path) {
+                Ok(installation_date) => installation_date,
+                Err(e) => {
+                    log::warn!("[Nimbus] Unable to get installation date from path, defaulting to today: {:?}", e);
+                    Utc::now()
+                }
+            };
+            let store = db.get_store(StoreId::Meta);
+            store.put(writer, DB_KEY_INSTALLATION_DATE, &installation_date)?;
+            installation_date
+        };
+        Ok(res)
+    }
+
+    fn get_update_date(&self, db: &Database, writer: &mut Writer) -> Result<DateTime<Utc>> {
+        let store = db.get_store(StoreId::Meta);
+
+        let persisted_app_version: Option<String> = store.get(writer, DB_KEY_APP_VERSION)?;
+        let res = match (persisted_app_version, &self.app_context.app_version) {
+            (persisted_app_version, Some(current_app_version))
+                if (persisted_app_version.is_some()
+                    && persisted_app_version.clone().unwrap() != *current_app_version)
+                    || persisted_app_version.is_none() =>
+            {
+                store.put(writer, DB_KEY_APP_VERSION, current_app_version)?;
+                let update_date = Utc::now();
+                store.put(writer, DB_KEY_UPDATE_DATE, &update_date)?;
+                update_date
+            }
+            (Some(persisted), Some(current)) if persisted == *current => {
+                let update_date: Option<DateTime<Utc>> = store.get(writer, DB_KEY_UPDATE_DATE)?;
+                if let Some(update_date) = update_date {
+                    update_date
+                } else {
+                    let update_date = Utc::now();
+                    store.put(writer, DB_KEY_UPDATE_DATE, &update_date)?;
+                    update_date
+                }
+            }
+            _ => {
+                let update_date = Utc::now();
+                store.put(writer, DB_KEY_UPDATE_DATE, &update_date)?;
+                update_date
+            }
+        };
+        Ok(res)
```

Comment:
I think this could benefit from making the `match` a bit more explicit:

```suggestion
        let persisted_app_version: Option<String> = store.get(writer, DB_KEY_APP_VERSION)?;
        let update_date: Option<DateTime<Utc>> = store.get(writer, DB_KEY_UPDATE_DATE)?;

        Ok(match (persisted_app_version, &self.app_context.app_version, update_date) {
            // The app been run before, but has not just been updated.
            (Some(persisted), Some(current), Some(date)) if persisted == *current => {
                date
            }
            // The app has been run before, and just been updated.
            (Some(persisted), Some(current), _) if persisted != *current => {
                let now = Utc::now();
                store.put(writer, DB_KEY_APP_VERSION, current)?;
                store.put(writer, DB_KEY_UPDATE_DATE, &now)?;
                now
            }
            // The app has just been installed
            (None, Some(current), _) => {
                let now = Utc::now();
                store.put(writer, DB_KEY_APP_VERSION, current)?;
                store.put(writer, DB_KEY_UPDATE_DATE, &now)?;
                now
            }
            // The current version is not available, or the persisted date is not available.
            (_, _, Some(date)) => {
                date
            }
            // Either way, this doesn't appear to be a good production environment.
            _ => {
                Utc::now()
            }
        })
```